streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -796,23 +796,23 @@ public synchronized boolean close(final long timeout, final TimeUnit timeUnit) {
         } else {
             stateDirCleaner.shutdownNow();
 
-            // notify all the threads to stop; avoid deadlocks by stopping any
-            // further state reports from the thread since we're shutting down
-            for (final StreamThread thread : threads) {
-                thread.setStateListener(null);
-                thread.shutdown();
-            }
-            if (globalStreamThread != null) {
-                globalStreamThread.setStateListener(null);
-                globalStreamThread.shutdown();
-            }
-
             // wait for all threads to join in a separate thread;
             // save the current thread so that if it is a stream thread
             // we don't attempt to join it and cause a deadlock
             final Thread shutdownThread = new Thread(new Runnable() {
                 @Override
                 public void run() {
+                    // notify all the threads to stop; avoid deadlocks by stopping any
```

Comment:
Is it necessary to move this logic within the close thread? Maybe I'm still missing something..

Reply:
My thinking was that since ThreadState.shutdown() can now call ThreadState.completeShutdown(), its now possible that ThreadState.shutdown() will block on the producer exiting (or something else exiting... it seems to be closing a lot of stuff). If this wait is indeterminately long, then we need to call shutdown() from the close thread to make sure we honor the timeout passed in to close().

Reply:
Ack, that makes sense.